long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -18,62 +18,96 @@
+import org.apache.flink.streaming.api.functions.sink.PrintSinkFunction;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.streaming.api.functions.source.SourceFunction;
 import org.apache.flink.training.exercises.common.datatypes.TaxiRide;
 import org.apache.flink.training.exercises.common.sources.TaxiRideGenerator;
-import org.apache.flink.training.exercises.common.utils.ExerciseBase;
 import org.apache.flink.training.exercises.common.utils.MissingSolutionException;
 import org.apache.flink.util.Collector;
 
+import java.time.Duration;
+
 /**
- * The "Long Ride Alerts" exercise of the Flink training in the docs.
+ * The "Long Ride Alerts" exercise.
+ *
+ * <p>The goal for this exercise is to emit the rideIds for taxi rides with a duration of more than
+ * two hours. You should assume that TaxiRide events can be lost, but there are no duplicates.
  *
- * <p>The goal for this exercise is to emit START events for taxi rides that have not been matched
- * by an END event during the first 2 hours of the ride.
+ * <p>You should eventually clear any state you create.
  */
-public class LongRidesExercise extends ExerciseBase {
+public class LongRidesExercise {
+    private SourceFunction<TaxiRide> source;
+    private SinkFunction<Long> sink;
+
+    /** Creates a job using the source and sink provided. */
+    public LongRidesExercise(SourceFunction<TaxiRide> source, SinkFunction<Long> sink) {
+        this.source = source;
+        this.sink = sink;
+    }
 
     /**
-     * Main method.
+     * Creates and executes the long rides pipeline.
      *
-     * @throws Exception which occurs during job execution.
+     * <p>@throws Exception which occurs during job execution.
      */
-    public static void main(String[] args) throws Exception {
+    public void execute() throws Exception {
 
         // set up streaming execution environment
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
-        env.setParallelism(ExerciseBase.parallelism);
 
         // start the data generator
-        DataStream<TaxiRide> rides = env.addSource(rideSourceOrTest(new TaxiRideGenerator()));
+        DataStream<TaxiRide> rides = env.addSource(source, TypeInformation.of(TaxiRide.class));
 
-        DataStream<TaxiRide> longRides =
-                rides.keyBy((TaxiRide ride) -> ride.rideId).process(new MatchFunction());
+        // the WatermarkStrategy specifies how to extract timestamps and generate watermarks
+        WatermarkStrategy<TaxiRide> watermarkStrategy =
+                WatermarkStrategy.<TaxiRide>forBoundedOutOfOrderness(Duration.ofSeconds(60))
+                        .withTimestampAssigner((ride, timestamp) -> ride.getEventTime());
 
-        printOrTest(longRides);
+        // create the pipeline
+        rides.assignTimestampsAndWatermarks(watermarkStrategy)
+                .keyBy((TaxiRide ride) -> ride.rideId)
```

Comment:
Simpler? Or did you want to make the type explicit?
```suggestion
                .keyBy(ride -> ride.rideId)
```